src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp

```diff
@@ -113,30 +135,73 @@ IMergingAlgorithm::Status ReplacingSortedAlgorithm::merge()
 
             /// Write the data for the previous primary key.
             if (!selected_row.empty())
+                insertRow();
+
+            selected_row.clear();
+        }
+
+        if (current->isFirst()
+            && key_differs
+            && is_deleted_column_number == -1 /// Ignore optimization if we need to filter deleted rows.
+            && sources_origin_merge_tree_part_level[current->order] > 0
+            && !skipLastRowFor(current->order) /// Ignore optimization if last row should be skipped.
+            && (queue.size() == 1 || (queue.size() >= 2 && current.totallyLess(queue.nextChild()))))
+        {
+            /// This is special optimization if current cursor is totally less than next cursor
+            /// and current chunk has no duplicates (we assume that parts with non-zero level have no duplicates)
+            /// We want to insert current cursor chunk directly in merged data.
+
+            /// First if merged_data is not empty we need to flush it.
+            /// We will get into the same condition on next merge call.
+            if (merged_data->mergedRows() != 0)
+                return Status(merged_data->pull());
+
+            size_t source_num = current->order;
+            auto current_chunk = std::move(*sources[source_num].chunk);
+            size_t chunk_num_rows = current_chunk.getNumRows();
+
+            /// We will get the next block from the corresponding source, if there is one.
+            queue.removeTop();
+
+            if (enable_vertical_final)
             {
-                if (is_deleted_column_number!=-1)
-                {
-                    if (!(cleanup && assert_cast<const ColumnUInt8 &>(*(*selected_row.all_columns)[is_deleted_column_number]).getData()[selected_row.row_num]))
-                        insertRow();
-                }
-                else
-                    insertRow();
-                /// insertRow() may has not been called
-                saveChunkForSkippingFinalFromSelectedRow();
+                auto replace_final_selection = ColumnUInt64::create(chunk_num_rows);
+                auto & replace_final_data = replace_final_selection->getData();
+
+                std::iota(replace_final_data.begin(), replace_final_data.end(), 0);
+                current_chunk.getChunkInfos().add(std::make_shared<ChunkSelectFinalIndices>(std::move(replace_final_selection)));
```

Comment:
i understand that it is the same in current master, but looks wasteful. could we handle this case separately w/o creating a column explicitly?

Reply:
Yes, sure.